src/Nulls.jl

```diff
@@ -115,9 +115,170 @@ xor(::Integer, ::Null) = null
 *(d::Null, x::AbstractString) = null
 *(d::AbstractString, x::Null) = null
 
-replace(itr, x) = (ifelse(v !== null, v, x) for v in itr)
-skip(itr) = (v for v in itr if v !== null)
-fail(itr) = (v !== null ? v : throw(NullException()) for v in itr)
+# Iterators
+"""
+    Nulls.replace(itr, replacement)
+
+Return an iterator wrapping iterable `itr` which replaces [`null`](@ref) values with
+`replacement`. When applicable, the size of `itr` is preserved.
+
+See also: [`Nulls.skip`](@ref), [`Nulls.fail`](@ref)
+
+# Examples
+```jldoctest
+julia> collect(Nulls.replace([1, null, 2], 0))
+3-element Array{Int64,1}:
+ 1
+ 0
+ 2
+
+julia> collect(Nulls.replace([1 null; 2 null], 0))
+2×2 Array{Int64,2}:
+ 1  0
+ 2  0
+
+```
+"""
+replace(itr, replacement) = EachReplaceNull(itr, replacement)
+struct EachReplaceNull{T, U}
+    x::T
+    replacement::U
+end
+Base.iteratorsize(::Type{<:EachReplaceNull{T}}) where {T} =
+    Base.iteratorsize(T)
+Base.iteratoreltype(::Type{<:EachReplaceNull{T}}) where {T} =
+    Base.iteratoreltype(T)
+Base.length(itr::EachReplaceNull) = length(itr.x)
+Base.size(itr::EachReplaceNull) = size(itr.x)
+Base.start(itr::EachReplaceNull) = start(itr.x)
+Base.done(itr::EachReplaceNull, state) = done(itr.x, state)
+Base.eltype(itr::EachReplaceNull) =
+    Union{Nulls.T(eltype(itr.x)), typeof(itr.replacement)}
```

Comment:
If the user is replacing `null` values, presumably they would expect the resulting element type to be `typejoin(Nulls.T(eltype(itr.x)), typeof(itr.replacement))`. Otherwise, wouldn't `replace([1,null,3], 2.0)` have element type `Union{Int, Float64}`? Seems like this might cause some slowdowns.

Reply:
Hmm, that's a subtle distinction and I never know which one is the most appropriate. Wouldn't `typejoin` always return an abstract type when replacement is of a different type? I'm not sure `Real` is better than `Union{Int, Float64}`. A third option is to use promotion to choose the best type, and perform conversion on the fly. I'm really not sure what's the best approach, maybe in part because I don't have a use case in mind.

Reply:
Oh sorry, not `typejoin`, I mean promote types.

Reply:
OK. Makes sense to use `promote` indeed. I'll do that unless somebody is of a different opinion.